File: TeamCode/src/main/java/org/firstinspires/ftc/teamcode/math/MathUtil.java
package org.firstinspires.ftc.teamcode.math;

import static java.lang.Math.PI;

public class MathUtil {
    public static final double EPSILON = 1e-6;

    public static boolean approxEquals(double d1, double d2) {
        if (Double.isInfinite(d1)) {
            // Infinity - infinity is NaN, so we need a special case
            return d1 == d2;
        } else {
            return Math.abs(d1 - d2) < EPSILON;
        }
    }

    public static double angleWrap(double angle)
    {
        while (angle>PI) angle -= PI*2;
        while (angle<-PI) angle += PI*2;
        return angle;
    }

    public static double cosFromSin(double sin, double angle)
    {
        double cos = Math.sqrt(1-sin*sin);
        angle = angle % (PI*2);
        if ((angle>PI) && (angle<PI*2))
            cos *= -1;
        return cos;
    }
}

File: TeamCode/src/main/java/org/firstinspires/ftc/teamcode/math/MathUtil.kt
package org.firstinspires.ftc.teamcode.math

import kotlin.math.PI
import kotlin.math.abs
import kotlin.math.sqrt

object MathUtil {

    private const val EPSILON = 1e-6

    @JvmStatic
    @JvmOverloads
    fun approxEquals(d1: Double, d2: Double, minDelta: Double = EPSILON) = abs(d1 - d2) < minDelta

    @JvmStatic
    fun angleWrap(angle: Double): Double {
        var correctedAngle = angle
        while (correctedAngle > PI) correctedAngle -= PI * 2
        while (correctedAngle < -PI) correctedAngle += PI * 2
        return correctedAngle
    }

    @JvmStatic
    fun angleAverage(angle1: Double, angle2: Double) = angleWrap(angle1 + angleWrap(angle2 - angle1) / 2.0)

    @JvmStatic
    fun minAbs(a: Double, b: Double) = if (abs(a) < abs(b)) a else b

    @JvmStatic
    fun cosFromSin(sin: Double, angle: Double): Double {
        val correctedAngle = angleWrap(angle)
        var cos = sqrt(1 - sin * sin)
        if (correctedAngle > PI / 2 || correctedAngle < -PI / 2) cos *= -1.0
        return cos
    }
}